src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java

```diff
@@ -0,0 +1,471 @@
+ * once.</li>
+ *
+ * </ol>
+ *
+ * <p>This functional approach is less common with Java 7 because it requires create lots of
+ * anonymous implementations and is therefore more verbose. However, the same approach is far more
+ * common in Java 8 and using lambdas requires very little code. This will make it easy to convert
+ * to Java 8 in the near future.
+ *
+ * <h2>Packaging and Deploying</h2> The dialect framework uses Java's {@link
+ * java.util.ServiceLoader} mechanism to find and automatically register all {@link
+ * DatabaseDialectProvider} implementations on the classpath. Don't forget to include in your JAR
+ * file a {@code META-INF/services/io.confluent.connect.jdbc.dialect .DatabaseDialectProvider} file
+ * that contains the fully-qualified name of your implementation class (or one class per line if
+ * providing multiple implementations).
+ *
+ * <p>If the Dialect implementation is included in this project, the class and the service provider
+ * file will automatically be included in the JDBC Source connector's JAR file. However, it is also
+ * possible to create an implementation outside of this project and to package the dialect (s) and
+ * service provider file in a JAR file, and to then include that JAR file inside the JDBC
+ * Connector's plugin directory within a Connect installation.
+ *
+ * <p>This discovery and registration process uses DEBUG messages to report the {@link
+ * DatabaseDialectProvider} classes that are found and registered. If you have difficulties getting
+ * the connector to find and register your dialect implementation classes, check that your JARs have
+ * the service provider file and your JAR is included in the JDBC connector's plugin directory.
+ */
+public interface DatabaseDialect extends ConnectionProvider {
+
+  /**
+   * Create a new prepared statement using the specified database connection.
+   *
+   * @param db    the database connection; may not be null
+   * @param query the query expression for the prepared statement; may not be null
+   * @return a new prepared statement; never null
+   * @throws SQLException if there is an error with the database connection
+   */
+  PreparedStatement createPreparedStatement(
+      Connection db,
+      String query
+  ) throws SQLException;
+
+  /**
+   * Parse the supplied simple name or fully qualified name for a table into a {@link TableId}.
+   *
+   * @param fqn the fully qualified string representation; may not be null
+   * @return the table identifier; never null
+   */
+  TableId parseTableIdentifier(String fqn);
+
+  /**
+   * Get the identifier rules for this database.
+   *
+   * @return the identifier rules
+   */
+  IdentifierRules identifierRules();
+
+  /**
+   * Get a new {@link ExpressionBuilder} that can be used to build expressions with quoted
+   * identifiers.
+   *
+   * @return the builder; never null
+   * @see #identifierRules()
+   * @see IdentifierRules#expressionBuilder()
+   */
+  ExpressionBuilder expressionBuilder();
```

Comment:
There is quite a bit of code that builds SQL query and update expressions, and all of those need to be able to deal with fully-qualified identifiers _and_ the unique quoting and identifier rules. The `ExpressionBuilder` makes it easier to build a SQL expression without having to worry about those details, since the expression builder delegates to an `IdentifierRules` for the details. However, a DBMS can customize the behavior of `ExpressionBuilder`, and so the dialect must be able to create instances.